runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -0,0 +1,436 @@
+
+    protected Map<F, Multimap<I, Map.Entry<K,V>>> indexMap;
+
+    /** Generate a table with the given set of indexes. */
+    public CorfuTable(Class<F> indexFunctionEnumClass) {
+        indexerClass = indexFunctionEnumClass;
+        indexMap = new HashMap<>();
+        indexFunctions = EnumSet.allOf(indexFunctionEnumClass);
+        indexFunctions.forEach(f -> indexMap.put(f, ArrayListMultimap.create()));
+    }
+
+    /** Default constructor. Generates a table without any secondary indexes. */
+    public CorfuTable() {
+        indexFunctions = Collections.emptySet();
+        indexMap = Collections.emptyMap();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public int size() {
+        return mainMap.size();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public boolean isEmpty() {
+        return mainMap.isEmpty();
+    }
+
+    @Accessor
+    public boolean hasIndices() {return !indexFunctions.isEmpty();}
+
+    @Getter
+    public Class indexerClass;
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public boolean containsKey(@ConflictParameter Object key) {
+        return mainMap.containsKey(key);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public boolean containsValue(Object value) {
+        return mainMap.containsValue(value);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Accessor
+    public V get(@ConflictParameter Object key) {
+        return mainMap.get(key);
+    }
+
+
+
+    /** Get a mapping using the specified index function.
+     *
+     * @param indexFunction The index function to use.
+     * @param index         The index
+     * @return
+     */
+    @SuppressWarnings("unchecked")
+    public Collection<Object> getByIndex(@Nonnull F indexFunction, I index) {
+        return (Collection<Object>) indexFunction.getProjectionFunction()
+                                .generateProjection(index, indexMap
+                                .get(indexFunction).get(index).parallelStream())
+                                .collect(Collectors.toList());
+    }
+
+    /**
+     * Register new index class
+     *
+     * This replaces the current index.
+     *
+     * @param indexFunctionEnumClass
+     */
+    public void registerIndex(Class indexFunctionEnumClass) {
+        indexerClass = indexFunctionEnumClass;
+        indexMap = new HashMap<>();
+        indexFunctions = EnumSet.allOf(indexFunctionEnumClass);
+        indexFunctions.forEach(f -> indexMap.put(f, ArrayListMultimap.create()));
+        mainMap.forEach(this::mapSecondaryIndexes);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @MutatorAccessor(name = "put", undoFunction = "undoPut", undoRecordFunction = "undoPutRecord")
+    public V put(@ConflictParameter K key, V value) {
+        V previous = mainMap.put(key, value);
+        // If we have index functions, update the secondary indexes.
+        if (indexFunctions.size() > 0) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Use isEmpty() to check whether the collection is empty or not. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1155)